keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -327,17 +320,67 @@ like the `all` macro, e.g. `self.all(listItem, <predicate>)` or `self.all(mapKey
+enum: ["low", "medium", "high"]
+x-kubernetes-validations:
+- rule: "!(self == 'high' && oldSelf == 'low') && !(self == 'low' && oldSelf == 'high')"
+  message: cannot transition directly between 'low' and 'high'
+```
+
+Unlike other rules, transition rules apply only to operations meeting the following criteria:
+
+- Both an old and a new value exist. It remains possible to check if a value has been added or
+  removed by placing a transition rule on the parent node. Transition rules are never applied to
+  custom resource creation. When placed on an optional field, a transition rule will not apply to
+  update operations that set or unset the field.
+
+- The path to the schema node being validated by a transition rule must resolve to a node that is
+  comparable between the old object and the new object. For example, list items and their
+  descendants (`spec.foo[10].bar`) can't necessarily be correlated between an existing object and a
+  later update to the same object.
+
+  - Semantics from [server-side
+    apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) will be
+    honored. In particular, updates to descendants of collection types that are mergeable according
+    to server-side apply may be validated by transition rules. This includes the elements of maps
+    marked `x-kubernetes-map-type=granular` and lists marked
+    `x-kubernetes-list-type=map`. Transition rules apply to elements of `map`-type lists only when
+    an element exists in both the old and new object having identical values for all key
+    fields. Elements of lists marked `x-kubernetes-list-type=set` and their descendants do not
+    support transition rules, however, set membership changes are accessible to transition rules
+    defined on the parent (i.e. array) node.
+
+  - Errors will be generated by CRD writes if a schema node contains a transition rule that can
+    never be applied, e.g. "<path>: update rule <rule> cannot be set on schema because the schema or
+    its parent schema is not mergeable".
+
+If all of the above criteria are satisfied for a given operation, the transition rule will be
+enforced. The identifier `oldSelf` is guaranteed to be bound to a non-null value during expression
+evaluation.
+
+It is possible for an attempt to update a custom resource to be rejected solely due to a disallowed
+transition from existing state. This [complicates declarative reconciliation
+tools](https://docs.google.com/document/d/1ZpCHE4yrOXoawai8Ldz49A4t9ni-QsHSuoy71VhBTLw/edit) that
+need to decide whether or not to delete-and-recreate an existing object. This problem exists today
+with fields that are immutable or immutable-once-assigned. Once a mechanism exists to communicate to
+clients that an operation was rejected due to transition errors only, that mechanism will be adopted
+for errors produced by the transition rules described in this document (see
+https://github.com/kubernetes/kubernetes/issues/107919).
```

Comment:
suggest moving this bit to the "design considerations" section and overwriting the `Possible interactions with [immutability and validation]` bit